File: app/graphql/mutations/create_division.rb
class Mutations::CreateDivision < Mutations::BaseMutation
  graphql_name "CreateDivision"

  argument :input, Inputs::CreateDivisionInput, required: true

  field :division, Types::Division, null: true
  field :success, Boolean, null: false
  field :message, String, null: true
  field :userErrors, [Types::Error], null: true
end

File: clients/admin/src/mutations/CreateDivision.ts
import client from "../modules/apollo";
import mutationPromise from "../helpers/mutationPromise";
import gql from "graphql-tag";

const mutation = gql`
  mutation CreateDivisionMutation($input: CreateDivisionInput!) {
    createDivision(input:$input) {
      division {
        id
        name
        numTeams
        numDays
        bracket {
          handle
        }
      }
      success
      message
      userErrors {
        field
        message
      }
    }
  }
`;

const update = () => {
  client.resetStore();
}

function commit(variables: CreateDivisionMutationVariables) {
  return mutationPromise((resolve, reject) => {
    client.mutate({
      mutation,
      variables,
      update
    }).then(({ data: { createDivision } }) => {
      resolve(createDivision as MutationResult);
    }).catch((error) => {
      reject(error);
    });
  });
}

export default { commit };
